doc/system_requirements.md

```diff
@@ -0,0 +1,265 @@
+# System Requirement Specification -- Exasol SQL Statement Builder
+
+## Introduction
+
+The Exasol SQL Statement Builder (ESB) is a [Java](https://java.com) library that allows you to define SQL statements in a [internal Domain-specific language (DSL)](https://en.wikipedia.org/wiki/Domain-specific_language#Usage_patterns). This means it uses standard Java language features to create a DSL.
+
+The project uses agile software development process, so this document contains only the portion of requirement necessary for the current iteration.
+
+### Goals
+
+The goals of the ESB are:
+
+* Abstraction from the SQL text representation
+* Compile-time error detection (where possible)
+* Extensibility (for different SQL dialects, validators and renderers)
+* User friendly API
+
+### Terms and Abbreviations
+
+* ESB: Exasol SQL Statement Builder
+* Renderer: extension that turns the abstract representation into a different one (most notably an SQL string)
+* Validator: extension that validates a statements structure and content
+
+### Notation
+
+#### Augmented Backus–Naur Form (ABNF)
+
+This document uses Augmented Backus–Naur Form (ABNF) for syntax definitions.
+
+#### ABNF Terminals
+
+This subsection list the ABNF terminals used in this document. Terminals are ABNF rules that cannot be split further down, like string literals for example.
+
+##### General Terminals
+
+    COMMA = ","
+    
+    L-BRACKET = "("
+    
+    R-BRACKET = ")"
+
+##### Operator Terminals
+
+    EQUAL-OPERATOR = "="
+    
+    NOT-EQUAL-OPERATOR = "<>"
+    
+    LESS-THAN-OPERATOR = "<"
+    
+    LESS-THAN-OR-EQUAL-OPERATOR = "<="
+    
+    GREATER-THAN-OPERATOR = ">"
+    
+    GREATER-THAN-OR-EQUAL-OPERATOR = ">="
+
+## Features
+
+### Statement Definition
+`feat~statement-definition~1`
+
+The ESB allows users to define SQL statements in abstract form.
+
+Needs: req
+
+### SQL String Rendering
+`feat~sql-string-rendering~1`
+
+The ESB renders abstract SQL statements into strings.
+
+Rationale:
+
+The SQL strings are necessary input for executing queries (e.g. with JDBC).
+
+Needs: req
+
+### Compile-time Error Checking
+`feat~compile-time-error-checking~1`
+
+ESB reports detectable errors at compile-time.
+
+Rationale:
+
+Making sure at compile time that illegal constructs do not compile make the resulting code safer and reduces debugging efforts.
+
+Needs: req
+
+## Functional Requirements
+
+### Statement Structure
+
+#### Building the Statement Structure Step-wise
+`req~statement-structure.step-wise~1`
+
+ESB lets users build the statement structure step-by-step.
+
+Rationale:
+
+This is necessary since complex statements are usually build as a result of multi-layered decision trees and parts of the statements are constructed in different places.
+
```

Comment:
How will those complex statements be built in ESB? 

(e.g)
```
 SqlStatement statement = StatementFactory.getInstance()
  .select().field("firstname", "lastname")
  .from(
 StatementFactory.getInstance()
  .select().field("foo", "bar")
  .from().table("fobar")
).table("person");
```
?

Reply:
Short term: pass the `SELECT` object handle and add the clauses in any order. This works in cases like e.g. the Informatica Connector.

Long term I will add overloaded methods like this

```java
public Select where(WhereClause whereClause)
```

and analogous for other clauses so that the can be constructed independently. For now the simple solutions is sufficient.